### PR TITLE
New documentation: cc65-intern.sgml

### PR DESCRIPTION
**Pull request by bbbradsmith**
Creating cc65-intern.sgml document to cover internal knowledge of cc65 for advanced users who wish to integrate with assembly code. This draft documents function call assembly protocol.

In response to this issue: https://github.com/cc65/cc65/issues/270

### CODE REVIEW

File: doc/cc65-intern.sgml [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+<sect1>Calling conventions<p>
+
+There are two calling conventions used in cc65:
+
+<itemize>
+  <item><tt/cdecl/ - passes all parameters on the C-stack.
+  <p>
+  <item><tt/fastcall/ - passes the rightmost parameter in
+  registers <tt>A/X/sreg</tt> an all others on the C-stack.
```
Comment (greg-king5) [inline doc/cc65-intern.sgml]: Typo: `an` --> `and`.

File: doc/cc65-intern.sgml [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+to <tt/int/ and be pushed as 2 bytes. K &amp; R style function prototypes may be used,
+but they will function the same as if no prototype was used.
+
+<sect1>Prologue, before the function call<p>
+
+If the function is declared as fastcall, the rightmost argument will be loaded into
+the <tt>A/X/sreg</tt> registers:
+
+<itemize>
+  <item><tt/A/ - 8-bit parameter, or low byte of larger tyes<p>
```
Comment (greg-king5) [inline doc/cc65-intern.sgml]: Typo: `tyes` --> `types`.

File: doc/cc65-intern.sgml [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+; Offset 0 ->+------------------+
+
+; Example code for accessing bar. The variable is in A/X after this code snippet:
+;
+    ldy     #2      ; Offset of high byte of bar
+    lda     (sp),y  ; High byte now in A
+    tax             ; High byte now in X
+    dey             ; Offset of low byte of bar
+    lda     (sp),y  ; Low byte now in A
+</verb></tscreen>
```
Comment (greg-king5) [inline doc/cc65-intern.sgml]: If `foo()` is `fastcall`, and is written in Assembly, then `baz` isn't put on the C-stack automatically.  That example needs to changed, in some way:

- Make `foo()` be `cdecl`.  Or,
- Push `baz` onto the C-stack.  Or,
- Store baz into a private variable.  (Then, change the offsets.)

Hmm, perhaps you could make two examples: one for `fastcall`, and another one for `cdecl`.

**Comment by bbbradsmith** [comment]
Updates pull request with those notes.

**Comment by oliverschmidt** [comment]
@bbbradsmith: Thanks for your contribution to the cc65@GitHub docs !

Please remove the corresponding section from the Wiki now that the content is the official docs.

**Pull request merged**